File: packages/models/src/Crag.ts
import Area, { type IArea } from "./Area.js";
import Bounds, { type IBounds } from "./Bounds.js";
import { type ICommentable } from "./Commentable.js";
import Coordinate, { type ICoordinate } from "./Coordinate.js";
import Parking, { type IParking } from "./Parking.js";
import { type IPhotoable } from "./Photoable.js";
import Trail, { type ITrail } from "./Trail.js";

export interface ICrag {
  id?: number; // IDs are tricky -- schema wise they always exist, but not till we save
  name: string;
  description?: string;

  bounds?: IBounds;
  center: ICoordinate;

  defaultZoom: number;
  minZoom?: number;
  maxZoom?: number;

  // Associations
  parking?: IParking;
  areas?: IArea[];
  commentable?: ICommentable;
  photoable?: IPhotoable;
  trail?: ITrail;
}

interface Crag extends Omit<ICrag, "center" | "parking"> {}
class Crag {
  bounds?: Bounds; // We map embedded entities with nullable columns to a nullable property.
  parking?: Parking;
  center: Coordinate;
  areas: Area[];
  trail?: Trail;

  static build(data: ICrag) {
    if (data instanceof Crag) return data;

    return new Crag(data);
  }

  constructor(data: ICrag) {
    this.id = data.id;
    this.name = data.name;
    this.description = data.description;

    this.bounds = Bounds.build(data.bounds);
    this.center = Coordinate.build(data.center);
    this.areas = (data.areas || []).map((area) => new Area(area));
    if (data.trail) {
      this.trail = new Trail(data.trail);
    }
    if (data.parking) {
      this.parking = new Parking(data.parking);
    }

    this.defaultZoom = data.defaultZoom;
    this.minZoom = data.minZoom;
    this.maxZoom = data.maxZoom;
  }
}

export default Crag;

File: go-server/models/crag.go
package models

type Crag struct {
	ID          *int         `json:"id,omitzero"`
	Name        string       `json:"name"`
	Description *string      `json:"description,omitzero"`
	Bounds      *Bounds      `json:"bounds,omitzero"`
	Center      Coordinate   `json:"center"`
	DefaultZoom int          `json:"defaultZoom"`
	MinZoom     *int         `json:"minZoom,omitzero"`
	MaxZoom     *int         `json:"maxZoom,omitzero"`
	Parking     *Parking     `json:"parking,omitzero"`
	Areas       []Area       `json:"areas,omitzero"`
	Commentable *Commentable `json:"commentable,omitzero"`
	Photoable   *Photoable   `json:"photoable,omitzero"`
	Trail       *Trail       `json:"trail,omitzero"`
}

func (c Crag) IsZero() bool {
	return c.ID == nil && c.Name == "" && c.Description == nil && c.Bounds == nil && c.Center.IsZero() &&
		c.DefaultZoom == 0 && c.MinZoom == nil && c.MaxZoom == nil && c.Parking == nil &&
		len(c.Areas) == 0 && c.Commentable == nil && c.Photoable == nil && c.Trail == nil
}
